### PR TITLE
Authz: node18 re-write

### PR DESCRIPTION
**Pull request by sauntimo**
> [!CAUTION]
> Do not merge until the correct point in the [Deployment Plan](https://oktawiki.atlassian.net/wiki/spaces/Bacca/pages/3103917188/Authorization+Extension+Upgrade+Deployment+Plan)

## ✏️ Changes
  
- Update the extension to run on node `18.16.0`
- Bump underlying hapi.js framework to `21.3.3` which involved a huge number of changes because the whole framework has been re-written from using callbacks to using async/await.
- Many other related dependency version bumps, to get compatibility between peer dependencies. There are some limitations based on which versions of what packages can be included from https://auth0-extensions.github.io/canirequire/ - some specific package versions are not available and so are included in the bundle instead.
- fix functionality broken by package updates, which, essentially, was all of it
- fix all unit (144) and integration (34) tests. NB, integration tests are run pointing at a layer0 dev space.

## NOTES
- will only build with this UNMERGED branch of the auth0-extensions-cli: https://github.com/auth0-extensions/auth0-extensions-cli/pull/64
- check this branch out locally, then in the repo run `npm link`. Then navigate back to the authorization extension repo and run `npm link auth0-extensions-cli`. This will then use the _local_ version of the extensions cli, not the master version.
  
## 📷 Screenshots
 
If there were visual changes to the application with this change, please include before and after screenshots here. If it has animation, please use screen capture software like  to make a gif.
  
## 🔗 References
  
- Jira Epic: https://auth0team.atlassian.net/browse/IDS-4386
  
## 🎯 Testing

- The plan to test installing and updating the extension is [here](https://oktawiki.atlassian.net/wiki/spaces/Bacca/pages/3103917188/Authorization+Extension+Upgrade+Deployment+Plan#Test-Process%3A-Updating-from-node12-version-to-node18-version)  

✅ This change has been tested in a Webtask
 
✅ This change has unit test coverage
  
✅ This change has integration test coverage
  
✅ This change has been tested for performance
  
## 🚀 Deployment
  
> Can this change be merged at any time? What will the deployment of the change look like? Does this need to be released in lockstep with something else?
  
🚫 This should only be merged and deployed using the extensions deployment tool at the correct point in the [Deployment Plan](https://oktawiki.atlassian.net/wiki/spaces/Bacca/pages/3103917188/Authorization+Extension+Upgrade+Deployment+Plan)
  
## 🎡 Rollout
    
In order to verify that the deployment was successful we will install and update the extension in a prod env and check functionality.
  
## 🔥 Rollback
    
If there are issues with the extension, then will will update the `extensions.json` to the previous node12  version `2.11.0` of the authz extension and advise customers who have updated, to "update" again to the previous version.

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -31,11 +31,12 @@
         "SwitchCase": 1
       }
     ],
-    "import/no-extraneous-dependencies": [error, { "devDependencies": true }],
+    "import/no-extraneous-dependencies": ["error", { "devDependencies": true }],
     "import/namespace": 0,
     "import/named": 0,
     "import/prefer-default-export": 0,
     "import/default": 0,
+    "import/resolver": { "node": { "extensions": [".js",".jsx"] } },
```
Comment (fbgoode) [inline .eslintrc:39]: According to my IDE this config for import/resolver is invalid. It's expecting 0, 1 or 2.

Comment (sauntimo) [reply .eslintrc]: should be fixed in [170a9c1](https://github.com/auth0/auth0-authorization-extension/pull/387/commits/170a9c1d068bd9e5a13e2650b4b92b17a4aaf255) 😄

File: server/api/groups/routes/get_group.js [hunk 2]
```diff
@@ -1,35 +1,38 @@
...
+        return h.response(groupExpanded);
+      }
 
-    return req.storage.getGroup(req.params.id)
-      .then(group => reply({ _id: group._id, name: group.name, description: group.description }))
-      .catch(err => reply.error(err));
+      const group = await req.storage.getGroup(req.params.id);
+      return h.response({ _id: group._id, name: group.name, description: group.description });
+    } catch (error) {
+      throw Boom.badRequest(`The record ${req.params.id} in groups does not exist.`);
```
Comment (fbgoode) [inline server/api/groups/routes/get_group.js:35]: I think you'll be able to remove this try/catch after the changes you're making

File: server/api/permissions/routes/get_permission.js [hunk 3]
```diff
@@ -1,27 +1,32 @@
...
+      const permission = await req.storage.getPermission(req.params.id);
+      return h.response({
         _id: permission._id,
         name: permission.name,
         description: permission.description
-      }))
-      .catch(err => reply.error(err))
+      });
+    } catch (error) {
+      throw Boom.badRequest(`The record ${req.params.id} in permissions does not exist.`);
```
Comment (fbgoode) [inline server/api/permissions/routes/get_permission.js:29]: same

File: server/api/roles/routes/get_role.js [hunk 4]
```diff
@@ -1,27 +1,33 @@
...
+
+      return h.response({
         _id: role._id,
         name: role.name,
         description: role.description
-      }))
-      .catch(err => reply.error(err))
+      });
+    } catch (error) {
+      throw Boom.badRequest(`The record ${req.params.id} in roles does not exist.`);
```
Comment (fbgoode) [inline server/api/roles/routes/get_role.js:30]: same

File: server/api/users/routes/get_user.js [hunk 5]
```diff
@@ -1,25 +1,31 @@
...
-    req.pre.auth0.users.get({ id: req.params.id })
-      .then(user => reply(user))
-      .catch(err => reply.error(err))
+  handler: async (req, h) => {
+    try {
+      const user = await req.pre.auth0.users.get({ id: req.params.id });
+      return h.response(user);
+    } catch (error) {
+      // failing to get a user throws, so catch and return a 400
+      throw Boom.badRequest();
```
Comment (fbgoode) [inline server/api/users/routes/get_user.js:28]: same

File: server/lib/logger.js [hunk 6]
```diff
@@ -1,15 +1,15 @@
...
-winston.emitErrs = true;
-
-const logger = new winston.Logger({
+const logger = new winston.createLogger({
+  // handleExceptions: true,
+  format: winston.format.json(),
   transports: [
     new winston.transports.Console({
-      timestamp: true,
+      // timestamp: true,
```
Comment (fbgoode) [inline server/lib/logger.js:8]: If those are now invalid options we can remove the comments?

File: server/plugins/auth.js [hunk 7]
```diff
@@ -61,92 +63,96 @@ export const register = (server, options, next) => {
...
-    register: tools.plugins.dashboardAdminSession,
+    plugin,
     options: {
       stateKey: 'authz-state',
       nonceKey: 'authz-nonce',
       sessionStorageKey: 'authz:apiToken',
       rta: config('AUTH0_RTA').replace('https://', ''),
       domain: config('AUTH0_DOMAIN'),
-      scopes: 'read:resource_servers create:resource_servers update:resource_servers delete:resource_servers read:clients read:connections read:rules create:rules update:rules update:rules_configs read:users',
+      // scopes: 'read:resource_servers create:resource_servers update:resource_servers delete:resource_servers read:clients read:connections read:rules create:rules update:rules update:rules_configs read:users',
```
Comment (fbgoode) [inline server/plugins/auth.js:136]: Remove comment?

Comment (sauntimo) [reply server/plugins/auth.js]: done in [46da9cc](https://github.com/auth0/auth0-authorization-extension/pull/387/commits/46da9cc5fb0dec70c6ea0e1fcb065ba66436c1e6)

File: server/lib/logger.js (hunk 6 above)
Comment (sauntimo) [reply server/lib/logger.js]: done in [46da9cc](https://github.com/auth0/auth0-authorization-extension/pull/387/commits/46da9cc5fb0dec70c6ea0e1fcb065ba66436c1e6)

**Review by fbgoode — DISMISSED** [review]

**Review by fbgoode — APPROVED** [review]
Tested changes succesfully in Layer0 dev space.

**Review by VictorGarridoAuth0 — APPROVED** [review]

**Pull request merged**